daemon/daemon.go

```diff
@@ -194,42 +192,27 @@ func (daemon *Daemon) UsesSnapshotter() bool {
 // RegistryHosts returns the registry hosts configuration for the host component
 // of a distribution image reference.
 func (daemon *Daemon) RegistryHosts(host string) ([]docker.RegistryHost, error) {
-	var (
-		conf        = daemon.config()
-		registryKey = "docker.io"
-		mirrors     = make([]string, len(conf.Mirrors))
-		m           = map[string]resolverconfig.RegistryConfig{}
-	)
-	// must trim "https://" or "http://" prefix
-	for i, v := range conf.Mirrors {
-		if uri, err := url.Parse(v); err == nil {
-			v = uri.Host
-		}
-		mirrors[i] = v
-	}
-	// set mirrors for default registry
-	m[registryKey] = resolverconfig.RegistryConfig{Mirrors: mirrors}
+	var m = map[string]resolverconfig.RegistryConfig{}
 
-	for _, v := range conf.InsecureRegistries {
-		u, err := url.Parse(v)
-		if err != nil && !strings.HasPrefix(v, "http://") && !strings.HasPrefix(v, "https://") {
-			originalErr := err
-			u, err = url.Parse("http://" + v)
-			if err != nil {
-				err = originalErr
-			}
-		}
+	mirrors := daemon.registryService.ServiceConfig().Mirrors
+	m["docker.io"] = resolverconfig.RegistryConfig{Mirrors: mirrors}
+
+	conf := daemon.registryService.ServiceConfig().IndexConfigs
+	for k, v := range conf {
 		c := resolverconfig.RegistryConfig{}
-		if err == nil {
-			v = u.Host
+		if !v.Secure {
 			t := true
-			if u.Scheme == "http" {
-				c.PlainHTTP = &t
-			} else {
-				c.Insecure = &t
-			}
+			c.PlainHTTP = &t
+			c.Insecure = &t
```

Comment:
Does this mean we assume _all_ insecure registries are plain HTTP, or is this used to trigger the "try HTTPS ignoring the cert, then fall back to HTTP" logic?  Sorry, I'm not very familiar with the underlying code here :sob:

(IMO this detection behavior of "insecure-registries" is still a misfeature and it'd be a benefit to us to go the other direction and encourage users to make it explicit, but I think I might've already lost that fight with some of the previous related PRs :see_no_evil:)

Reply:
> or is this used to trigger the "try HTTPS ignoring the cert, then fall back to HTTP" logic

Exactly this :tada: This is how we get here in BuildKit: https://github.com/moby/buildkit/blob/master/util/resolver/resolver.go#L25.

This function gets called in `resolver.NewRegistryConfig` which is called by moby, which returns the two hosts, which get tried in order (return on first success).

> and it'd be a benefit to us to go the other direction and encourage users to make it explicit, but I think I might've already lost that fight with some of the previous related PRs

I don't think this battle is necessarily lost -- given *our* lack of knowledge about how this worked, I think it's fair to assume most users don't really know. So we could make a point of deprecating the current behavior at some point, and try and require protocols.